gum/arch-mips/gummipsrelocator.c

```diff
@@ -0,0 +1,530 @@
+                                 mips_reg * available_scratch_reg)
+{
+  guint n = 0;
+  guint8 * buf;
+  GumMipsWriter cw;
+  GumMipsRelocator rl;
+  guint reloc_bytes;
+
+  buf = g_alloca (3 * min_bytes);
+  gum_mips_writer_init (&cw, buf);
+
+  gum_mips_relocator_init (&rl, address, &cw);
+
+  do
+  {
+    const cs_insn * insn;
+    gboolean safe_to_relocate_further;
+
+    reloc_bytes = gum_mips_relocator_read_one (&rl, &insn);
+    if (reloc_bytes == 0)
+      break;
+
+    n = reloc_bytes;
+
+    if (scenario == GUM_SCENARIO_ONLINE)
+    {
+      switch (insn->id)
+      {
+        case MIPS_INS_JAL:
+        case MIPS_INS_JALR:
+        case MIPS_INS_SYSCALL:
+          safe_to_relocate_further = FALSE;
+          break;
+        default:
+          safe_to_relocate_further = TRUE;
+          break;
+      }
+    }
+    else
+    {
+      safe_to_relocate_further = TRUE;
+    }
+
+    if (!safe_to_relocate_further)
+      break;
+  }
+  while (reloc_bytes < min_bytes || rl.delay_slot_pending);
+
+  if (!rl.eoi)
+  {
+    csh capstone;
+    cs_err err;
+    cs_insn * insn;
+    size_t count, i;
+    gboolean eoi;
+
+#ifdef G_LITTLE_ENDIAN
+    err = cs_open (CS_ARCH_MIPS, CS_MODE_MIPS32 | CS_MODE_LITTLE_ENDIAN,
+        &capstone);
+#else
+    err = cs_open (CS_ARCH_MIPS, CS_MODE_MIPS32 | CS_MODE_BIG_ENDIAN,
+        &capstone);
+#endif
+    g_assert_cmpint (err, == , CS_ERR_OK);
+    err = cs_option (capstone, CS_OPT_DETAIL, CS_OPT_ON);
+    g_assert_cmpint (err, ==, CS_ERR_OK);
+
+    count = cs_disasm (capstone, rl.input_cur, 1024, rl.input_pc, 0, &insn);
+    g_assert (insn != NULL);
+
+    eoi = FALSE;
+    for (i = 0; i != count && !eoi; i++)
+    {
+      cs_mips * d = &insn[i].detail->mips;
+
+      switch (insn[i].id)
+      {
+        case MIPS_INS_J:
+        {
+          cs_mips_op * op = &d->operands[0];
+          g_assert (op->type == MIPS_OP_IMM);
+          gssize target =
+            (gssize) (GPOINTER_TO_SIZE (insn[i].address & 0xf0000000)) |
+            (op->imm << 2);
+          gssize offset = target - (gssize) GPOINTER_TO_SIZE (address);
+          if (offset > 0 && offset < (gssize) n)
+            n = offset;
+          eoi = TRUE;
+          break;
+        }
+        case MIPS_INS_BEQ:
+        case MIPS_INS_BEQL:
+        case MIPS_INS_BGEZ:
+        case MIPS_INS_BGEZL:
+        case MIPS_INS_BGTZ:
+        case MIPS_INS_BGTZL:
+        case MIPS_INS_BLEZ:
+        case MIPS_INS_BLEZL:
+        case MIPS_INS_BLTZ:
+        case MIPS_INS_BLTZL:
+        case MIPS_INS_BNE:
+        case MIPS_INS_BNEL:
+        {
+          cs_mips_op * op =
+            d->op_count == 3 ? &d->operands[2] : &d->operands[1];
+          g_assert (op->type == MIPS_OP_IMM);
+          gssize target = (gssize) insn->address +
+            (op->imm & 0x8000 ? (0xffff0000 + op->imm) << 2 : op->imm << 2);
+          gssize offset =
+              target - (gssize) GPOINTER_TO_SIZE (address);
```

Comment:
We should avoid declaration after statement for consistency. This also applies to the code above.